### ISSUE TITLE
System.Text.Json 4.7.2 uses System.Buffers >= 4.5.1(4.0.3.0 dll version) but asks for System.Buffers, Version=4.0.2.0

### ISSUE THREAD

**Issue by isxaker** [dotnet/runtime#38782]
### Description

Hi, there.

I use `System.Text.Json 4.7.2` nuget package.
It has the following dependencies:
[https://www.nuget.org/packages/System.Text.Json/4.7.2](https://www.nuget.org/packages/System.Text.Json/4.7.2)
.NETFramework 4.6.1
Microsoft.Bcl.AsyncInterfaces (>= 1.1.0)
**System.Buffers (>= 4.5.1)**
System.Memory (>= 4.5.4)
System.Numerics.Vectors (>= 4.5.0)
System.Runtime.CompilerServices.Unsafe (>= 4.7.1)
System.Text.Encodings.Web (>= 4.7.1)
System.Threading.Tasks.Extensions (>= 4.5.4)
System.ValueTuple (>= 4.5.0)

It's said **System.Buffers (>= 4.5.1)**. Which is **4.0.3.0** version of the dll .

Ok. I've downloaded the nuget package. And open `lib\net461\System.Text.Json.dll` with `dotPeek`.
I found that in reality it uses **System.Buffers 4.0.2.0** dll.

Because of it I got the following error:
(I have .net framework 4.7.2 class library which is used in powershell.exe)

```
7/4/2020 3:57:28 PM   24 (102232) Error: Could not load file or assembly 'System.Buffers, Version=4.0.2.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' or one of its dependencies. The system cannot find the file specified.
7/4/2020 3:57:28 PM   24 (102232) Type: System.IO.FileNotFoundException
7/4/2020 3:57:28 PM   24 (102232) Stack:
7/4/2020 3:57:28 PM   24 (102232)    at System.Text.Json.PooledByteBufferWriter..ctor(Int32 initialCapacity)
   at System.Text.Json.JsonSerializer.WriteCoreString(Object value, Type type, JsonSerializerOptions options)
   at Veeam.StandBy.Service.JsonProtocol.Serialization.SystemTextJsonSerializer.Serialize[T](T dto)
   at Veeam.StandBy.Service.JsonProtocol.Serialization.RequestSerializer.SerializeRequest[T](T request)
```
   

Configuration:
.net framework class library
I've added **System.Text.Json 4.7.2** nuget package by package reference

**FUS LOG**

```
*** Assembly Binder Log Entry  (7/4/2020 @ 4:31:22 PM) ***

The operation failed.
Bind result: hr = 0x80070002. The system cannot find the file specified.

Assembly manager loaded from:  C:\Windows\Microsoft.NET\Framework64\v4.0.30319\clr.dll
Running under executable  C:\Windows\System32\WindowsPowerShell\v1.0\powershell_ise.exe
--- A detailed error log follows. 

=== Pre-bind state information ===
LOG: DisplayName = System.Buffers, Version=4.0.2.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51
 (Fully-specified)
LOG: Appbase = file:///C:/Windows/System32/WindowsPowerShell/v1.0/
LOG: Initial PrivatePath = NULL
LOG: Dynamic Base = NULL
LOG: Cache Base = NULL
LOG: AppName = powershell_ise.exe
Calling assembly : System.Text.Json, Version=4.0.1.2, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51.
===
LOG: This bind starts in LoadFrom load context.
WRN: Native image will not be probed in LoadFrom context. Native image will only be probed in default load context, like with Assembly.Load().
LOG: Using application configuration file: C:\Windows\System32\WindowsPowerShell\v1.0\powershell_ise.exe.Config
LOG: Using host configuration file: 
LOG: Using machine configuration file from C:\Windows\Microsoft.NET\Framework64\v4.0.30319\config\machine.config.
LOG: Post-policy reference: System.Buffers, Version=4.0.2.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51
LOG: GAC Lookup was unsuccessful.
LOG: Attempting download of new URL file:///C:/Windows/System32/WindowsPowerShell/v1.0/System.Buffers.DLL.
LOG: Attempting download of new URL file:///C:/Windows/System32/WindowsPowerShell/v1.0/System.Buffers/System.Buffers.DLL.
LOG: Attempting download of new URL file:///C:/Windows/System32/WindowsPowerShell/v1.0/System.Buffers.EXE.
LOG: Attempting download of new URL file:///C:/Windows/System32/WindowsPowerShell/v1.0/System.Buffers/System.Buffers.EXE.
LOG: Attempting download of new URL file:///C:/Users/mbryksin/Documents/Veeam/Workspaces/2/Current/x64/Debug/System.Buffers.DLL.
LOG: Assembly download was successful. Attempting setup of file: C:\Users\mbryksin\Documents\Veeam\Workspaces\2\Current\x64\Debug\System.Buffers.dll
LOG: Entering run-from-source setup phase.
LOG: Assembly Name is: System.Buffers, Version=4.0.3.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51
WRN: Comparing the assembly name resulted in the mismatch: Build Number
ERR: The assembly reference did not match the assembly definition found.
ERR: Run-from-source setup phase failed with hr = 0x80131040.
LOG: Attempting download of new URL file:///C:/Users/mbryksin/Documents/Veeam/Workspaces/2/Current/x64/Debug/System.Buffers/System.Buffers.DLL.
LOG: Attempting download of new URL file:///C:/Users/mbryksin/Documents/Veeam/Workspaces/2/Current/x64/Debug/System.Buffers.EXE.
LOG: Attempting download of new URL file:///C:/Users/mbryksin/Documents/Veeam/Workspaces/2/Current/x64/Debug/System.Buffers/System.Buffers.EXE.
LOG: All probing URLs attempted and failed.

*** Assembly Binder Log Entry  (7/4/2020 @ 4:31:22 PM) ***

The operation failed.
Bind result: hr = 0x80070002. The system cannot find the file specified.

Assembly manager loaded from:  C:\Windows\Microsoft.NET\Framework64\v4.0.30319\clr.dll
Running under executable  C:\Windows\System32\WindowsPowerShell\v1.0\powershell_ise.exe
--- A detailed error log follows. 

=== Pre-bind state information ===
LOG: DisplayName = System.Buffers, Version=4.0.2.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51
 (Fully-specified)
LOG: Appbase = file:///C:/Windows/System32/WindowsPowerShell/v1.0/
LOG: Initial PrivatePath = NULL
LOG: Dynamic Base = NULL
LOG: Cache Base = NULL
LOG: AppName = powershell_ise.exe
Calling assembly : System.Text.Json, Version=4.0.1.2, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51.
===
LOG: This bind starts in LoadFrom load context.
WRN: Native image will not be probed in LoadFrom context. Native image will only be probed in default load context, like with Assembly.Load().
LOG: Using application configuration file: C:\Windows\System32\WindowsPowerShell\v1.0\powershell_ise.exe.Config
LOG: Using host configuration file: 
LOG: Using machine configuration file from C:\Windows\Microsoft.NET\Framework64\v4.0.30319\config\machine.config.
LOG: Post-policy reference: System.Buffers, Version=4.0.2.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51
LOG: GAC Lookup was unsuccessful.
LOG: Attempting download of new URL file:///C:/Windows/System32/WindowsPowerShell/v1.0/System.Buffers.DLL.
LOG: Attempting download of new URL file:///C:/Windows/System32/WindowsPowerShell/v1.0/System.Buffers/System.Buffers.DLL.
LOG: Attempting download of new URL file:///C:/Windows/System32/WindowsPowerShell/v1.0/System.Buffers.EXE.
LOG: Attempting download of new URL file:///C:/Windows/System32/WindowsPowerShell/v1.0/System.Buffers/System.Buffers.EXE.
LOG: Attempting download of new URL file:///C:/Users/mbryksin/Documents/Veeam/Workspaces/2/Current/x64/Debug/System.Buffers.DLL.
LOG: Assembly download was successful. Attempting setup of file: C:\Users\mbryksin\Documents\Veeam\Workspaces\2\Current\x64\Debug\System.Buffers.dll
LOG: Entering run-from-source setup phase.
LOG: Assembly Name is: System.Buffers, Version=4.0.3.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51
WRN: Comparing the assembly name resulted in the mismatch: Build Number
ERR: The assembly reference did not match the assembly definition found.
ERR: Run-from-source setup phase failed with hr = 0x80131040.
LOG: Attempting download of new URL file:///C:/Users/mbryksin/Documents/Veeam/Workspaces/2/Current/x64/Debug/System.Buffers/System.Buffers.DLL.
LOG: Attempting download of new URL file:///C:/Users/mbryksin/Documents/Veeam/Workspaces/2/Current/x64/Debug/System.Buffers.EXE.
LOG: Attempting download of new URL file:///C:/Users/mbryksin/Documents/Veeam/Workspaces/2/Current/x64/Debug/System.Buffers/System.Buffers.EXE.
LOG: All probing URLs attempted and failed.

```

Note:
1. everything is ok if I use my dll from *.exe application.
2. I've tried biding redirect - the same issue
3. I've tried adding System.Text.Json nuget package directly - the same issue

**Comment by joperezr** [comment]
This is technically ok and the fix is to include a binding redirect. The reason why it works just fine when you consume it from an .exe is most likely because that application is getting a binding redirect which is fixing this. You mentioned you tried generating binding redirects but that didn't work for you, how did you try to generate them? Depending on the SDK you are using it may be different, but usually for class libraries what you need to set in terms of properties for binding redirects to get automatically generated is to add:

```xml
<PropertyGroup>
  <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
  <GenerateBindingRedirectsOutputType>true</GenerateBindingRedirectsOutputType>
</PropertyGroup>
```

I tried that locally and then loaded the assembly from powershell and it worked fine for me

**Comment by isxaker** [comment]
> This is technically ok and the fix is to include a binding redirect. The reason why it works just fine when you consume it from an .exe is most likely because that application is getting a binding redirect which is fixing this. You mentioned you tried generating binding redirects but that didn't work for you, how did you try to generate them? Depending on the SDK you are using it may be different, but usually for class libraries what you need to set in terms of properties for binding redirects to get automatically generated is to add:
> 
> ```
> <PropertyGroup>
>   <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
>   <GenerateBindingRedirectsOutputType>true</GenerateBindingRedirectsOutputType>
> </PropertyGroup>
> ```
> 
> I tried that locally and then loaded the assembly from powershell and it worked fine for me

Thanks for the idea. 
I'l try it and return back with feedback.

Just to double check it - you did get the same error and when you added binding redirects to .csproj file the error gone away, right ?

**Comment by isxaker** [comment]
> This is technically ok and the fix is to include a binding redirect. The reason why it works just fine when you consume it from an .exe is most likely because that application is getting a binding redirect which is fixing this. You mentioned you tried generating binding redirects but that didn't work for you, how did you try to generate them? Depending on the SDK you are using it may be different, but usually for class libraries what you need to set in terms of properties for binding redirects to get automatically generated is to add:
> 
> ```
> <PropertyGroup>
>   <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
>   <GenerateBindingRedirectsOutputType>true</GenerateBindingRedirectsOutputType>
> </PropertyGroup>
> ```
> 
> I tried that locally and then loaded the assembly from powershell and it worked fine for me

Could you please also share you test project ?
I've tried the same - add to my class library .csproj file the following lines:
```
<PropertyGroup>
  <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
  <GenerateBindingRedirectsOutputType>true</GenerateBindingRedirectsOutputType>
</PropertyGroup>
```

Then I imported that dll to ps.exe as usual but the error is still here.
Basically it didn't help me.

I've found only this way for redirecting .dlls in PS
[https://stackoverflow.com/a/33302645/364429](https://stackoverflow.com/a/33302645/364429)

**Comment by joperezr** [comment]
> Just to double check it - you did get the same error and when you added binding redirects to .csproj file the error gone away, right ?

Yes I do get the error as binding redirects are required in order to load that version of System.Text.Json package.

> Could you please also share you test project ?

Sure, I'm using SDK-style projects for the repro, here are the contents:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net461</TargetFramework>
    <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
    <GenerateBindingRedirectsOutputType>true</GenerateBindingRedirectsOutputType>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="System.Text.Json" Version="4.7.2" />
  </ItemGroup>

</Project>
```

The really important bit is that after you build this, you should see a `<yourProjectName>.dll.config` file dropped right next to your dll on the bin output folder with the binding redirects, for my example here are the contents of the one that got generated for me:

```xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <runtime>
    <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
      <dependentAssembly>
        <assemblyIdentity name="System.Buffers" publicKeyToken="cc7b13ffcd2ddd51" culture="neutral" />
        <bindingRedirect oldVersion="0.0.0.0-4.0.3.0" newVersion="4.0.3.0" />
      </dependentAssembly>
    </assemblyBinding>
    <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
      <dependentAssembly>
        <assemblyIdentity name="System.Runtime.CompilerServices.Unsafe" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
        <bindingRedirect oldVersion="0.0.0.0-4.0.6.0" newVersion="4.0.6.0" />
      </dependentAssembly>
    </assemblyBinding>
    <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
      <dependentAssembly>
        <assemblyIdentity name="System.Threading.Tasks.Extensions" publicKeyToken="cc7b13ffcd2ddd51" culture="neutral" />
        <bindingRedirect oldVersion="0.0.0.0-4.2.0.1" newVersion="4.2.0.1" />
      </dependentAssembly>
    </assemblyBinding>
    <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
      <dependentAssembly>
        <assemblyIdentity name="System.ValueTuple" publicKeyToken="cc7b13ffcd2ddd51" culture="neutral" />
        <bindingRedirect oldVersion="0.0.0.0-4.0.3.0" newVersion="4.0.3.0" />
      </dependentAssembly>
    </assemblyBinding>
  </runtime>
</configuration>
```

The important bit there to fix this problem you are hitting is that one with System.Buffers that will redirect the 4.0.2.0 request to 4.0.3.0 so you shouldn't need any assembly binding logic as at least locally I'm seeing that powershell is honoring the .dll.config file and correctly redirecting System.Buffers to the right version.

**Comment by isxaker** [comment]
@joperezr Let's add a bit of transitivity here.

I have a class library - ClassLibararyForPS. 
This ClassLibararyForPS does not contain System.Text.Json itself. 
But ClassLibararyForPS uses another class library - ClassLibrary_1. 
And ClassLibarry_1 has System.Text.Json which uses System.Buffers 4.0.3.0

Also ClassLibararyForPS uses one more ClassLibary_2 which uses System.Buffers 4.0.2.0

ClassLibararyForPS -> ClassLibrary_1 -> System.Text.Json -> System.Buffers 4.0.3.0
ClassLibararyForPS -> ClassLibary_2 -> System.Buffers 4.0.2.0

In bin folder I found ClassLibararyForPS.dll.config bug it does not contain any redirecting settings.

**Comment by joperezr** [comment]
If you are usihg SDK-style projects like the example I gave above for all ClassLibraryForPS, ClassLibrary_1, and ClassLibrary_2 then it should all just work. The reason is that when you use SDK style projects and PackageReference model, all of transitive dependencies will flow up to the consuming app, while if you are doing so via traditional MSBuild projects it won't so it is possible to hit issues like the ones you are hitting. Can you try converting your projects to all be SDK style and give it a try again to see if that fixes the issue?

**Comment by codingbloke** [comment]
I am having very similar problem with the System.ValueTuple reference.  The version included in the Nuget fetch for System.Text.Json is 4.0.3.0 however on the build server MSBUILD complaines with:

`D:\TFSBuild\Agent1\_work\11\s\code\XWeb\Web.metaproj : warning MSB3268: The primary reference "D:\TFSBuild\Agent1\_work\11\s\code\XWeb\WebBin\Release\MyCompany.System.dll" could not be resolved because it has an indirect dependency on the framework assembly "System.ValueTuple, Version=4.0.2.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51" which could not be resolved in the currently targeted framework. ".NETFramework,Version=v4.8". To resolve this problem, either remove the reference "D:\TFSBuild\Agent1\_work\11\s\code\XWeb\WebBin\Release\MyCompany.System.dll" or retarget your application to a framework version which contains "System.ValueTuple, Version=4.0.2.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51".`

Both the App.Config for the System dll above and the ye olde website web.config contain these:

```xml
      <dependentAssembly>
        <assemblyIdentity name="System.ValueTuple" publicKeyToken="cc7b13ffcd2ddd51" culture="neutral" />
        <bindingRedirect oldVersion="0.0.0.0-4.0.3.0" newVersion="4.0.3.0" />
      </dependentAssembly>
      <dependentAssembly>
        <assemblyIdentity name="System.Buffers" publicKeyToken="cc7b13ffcd2ddd51" culture="neutral" />
        <bindingRedirect oldVersion="0.0.0.0-4.0.3.0" newVersion="4.0.3.0" />
      </dependentAssembly>
```

As per the advice above (and everywhere else on the tinterweb I've looked) but MSBuild is having none of it.

**Comment by imckl** [comment]
Also facing similar problem.
In my scenario, I've built an server-client structure application.
```
CommonLibrary: .netstandard2.0 (System.Text.Json-4.7.2 from nuget)
Server: .NetCore3.1 for both core and wpf project (support System.Text.Json, internall)
Client: .NetStandard2.0 for core and .NetFramework 4.7.2 fot wpf project (System.Text.Json-4.7.2 from nuget)
```
When I use something like 
```<AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>``` 
and manually add 
```
      <dependentAssembly>
        <assemblyIdentity name="System.ValueTuple" publicKeyToken="cc7b13ffcd2ddd51" culture="neutral" />
        <bindingRedirect oldVersion="0.0.0.0-4.0.3.0" newVersion="4.0.3.0" />
      </dependentAssembly>
      <dependentAssembly>
        <assemblyIdentity name="System.Buffers" publicKeyToken="cc7b13ffcd2ddd51" culture="neutral" />
        <bindingRedirect oldVersion="0.0.0.0-4.0.3.0" newVersion="4.0.3.0" />
      </dependentAssembly>
```
, the client applications still throw same exception
```
System.IO.FileLoadException: 未能加载文件或程序集“System.Buffers, Version=4.0.2.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51”或它的某一个依赖项。找到的程序集清单定义与程序集引用不匹配。 (异常来自 HRESULT:0x80131040)
```

To reduce the complexity of the application, I've change all System.Text.Json code back to Newtonsoft.Json.
Hope there is a more proper way to resolve this issue

**Comment by ericstj** [comment]
@joperezr's suggestion of using bindingRedirects is the best solution here.  I tested the latest build of Json and it doesn't need bindingRedirects for the components listed (System.Buffers, System.ValueTuple, System.Threading.Tasks.Extensions) but it does still need a bindingRedirect for System.Runtime.CompilerServices.Unsafe.

I'm not seeing a great way to completely fix this given the dependency graph of these components.

System.Text.Json references System.Buffers, System.Memory, System.Numerics.Vectors, System.Threading.Tasks.Extensions, System.ValueTuple: all of which no longer build as packages in 5.0 since they were absorbed into .NETCore / .NETStandard.  

Some of these reference System.Runtime.CompilerServices.Unsafe, which will be the old version.  System.Runtime.CompilerServices.Unsafe still builds as a package in 5.0 and contains bugfixes over the old version referenced by those components.  I don't think we should be shipping Json with a reference to the old version of Unsafe, nor do I think we should be bringing back out-of-band builds of all these components into the latest codebase.  BindingRedirects are the unfortunate requirement of working on .NETFramework and using nuget packages.

**Comment by chucklu** [comment]
@isxaker Have you tried the solution here  [Powershell config assembly redirect](https://stackoverflow.com/a/33302645/13338936)?

**Comment by rjmholt** [comment]
I've also run into this trying to write a PowerShell 5.1 module, where System.Runtime.CompilerServices.Unsafe has the wrong version. I build a module against Azure.Bicep.Types.Az, which references System.Text.Json version 5.0.1. This works great in .NET 3.1+ (PowerShell 7), but in WinPS I get this:

```
System.Management.Automation.ParameterBindingException: Cannot retrieve the dynamic parameters for the cmdlet. Could not load file or assembly 'System.Text.Json, Version=5.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' or one of its dependencies. The system cannot find the file specified. ---> System.IO.FileNotFoundException: Could not load file or assembly 'System.Text.Json, Version=5.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' or one of its dependencies. The system cannot find the file specified.
   at Azure.Bicep.Types.Az.Index.TypeIndexer.DeserializeIndex(String content)
   at Azure.Bicep.Types.Az.TypeLoader.GetIndexedTypes()
   at PSArm.Schema.ResourceIndex.GetAllAvailableTypeLocations()
   at System.Lazy`1.CreateValue()
   at System.Lazy`1.LazyInitValue()
   at PSArm.Schema.ResourceIndex.get_AvailableTypeLocationList()
   at PSArm.Schema.ResourceIndex.LoadResourceSchemas()
   at System.Lazy`1.CreateValue()
   at System.Lazy`1.LazyInitValue()
   at PSArm.Schema.ResourceIndex.get_ResourceSchemaTable()
   at PSArm.Schema.ResourceIndex.TryGetResourceSchema(String providerNamespace, String providerName, String apiVersion, ResourceSchema& resourceSchema)
   at PSArm.Commands.Template.NewPSArmResourceCommand.TryGetResourceSchema(ResourceSchema& resourceSchema)
   at PSArm.Commands.Template.NewPSArmResourceCommand.GetDynamicParameters()
   at System.Management.Automation.CmdletParameterBinderController.HandleCommandLineDynamicParameters(ParameterBindingException& outgoingBindingException)
   --- End of inner exception stack trace ---
```

I tried explicitly importing the System.Text.Json DLL in my module before the main DLL is loaded, but I got the same exception.

Investigating this, I cut my own module out entirely and was able to proceed simply from importing the System.Text.Json assembly:

```powershell
> Add-Type -Path ./System.Text.Json.dll
> [System.Text.Json.JsonSerializer]::Deserialize('{}', [object])
Exception calling "Deserialize" with "2" argument(s): "Could not load file or assembly
'System.Runtime.CompilerServices.Unsafe, Version=4.0.4.1, Culture=neutral,
PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file
specified."
At line:1 char:1
+ [System.Text.Json.JsonSerializer]::Deserialize('{}', [object])
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : FileNotFoundException

> $error[0] |fl * -force


PSMessageDetails      :
Exception             : System.Management.Automation.MethodInvocationException: Exception calling
                        "Deserialize" with "2" argument(s): "Could not load file or assembly
                        'System.Runtime.CompilerServices.Unsafe, Version=4.0.4.1, Culture=neutral,
                        PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system
                        cannot find the file specified." ---> System.IO.FileNotFoundException: Could
                        not load file or assembly 'System.Runtime.CompilerServices.Unsafe,
                        Version=4.0.4.1, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of
                        its dependencies. The system cannot find the file specified.
                           at System.Span`1..ctor(T[] array)
                           at System.Text.Json.JsonSerializer.Deserialize[TValue](String json, Type
                        returnType, JsonSerializerOptions options)
                           at System.Text.Json.JsonSerializer.Deserialize(String json, Type
                        returnType, JsonSerializerOptions options)
                           at CallSite.Target(Closure , CallSite , Type , String , Type )
                           --- End of inner exception stack trace ---
                           at System.Management.Automation.ExceptionHandlingOps.ConvertToMethodInvoca
                        tionException(Exception exception, Type typeToThrow, String methodName,
                        Int32 numArgs, MemberInfo memberInfo)
                           at CallSite.Target(Closure , CallSite , Type , String , Type )
                           at
                        System.Dynamic.UpdateDelegates.UpdateAndExecute3[T0,T1,T2,TRet](CallSite
                        site, T0 arg0, T1 arg1, T2 arg2)
                           at System.Management.Automation.Interpreter.DynamicInstruction`4.Run(Inter
                        pretedFrame frame)
                           at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstructio
                        n.Run(InterpretedFrame frame)
TargetObject          :
CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
FullyQualifiedErrorId : FileNotFoundException
ErrorDetails          :
InvocationInfo        : System.Management.Automation.InvocationInfo
ScriptStackTrace      : at <ScriptBlock>, <No file>: line 1
PipelineIterationInfo : {}

> [system.span`1].Assembly.GetReferencedAssemblies()

Version        Name
-------        ----
4.0.0.0        mscorlib
4.0.0.0        System
4.1.4.0        System.Numerics.Vectors
4.0.4.1        System.Runtime.CompilerServices.Unsafe
4.0.3.0        System.Buffers

> [system.runtime.compilerservices.unsafe].Assembly.GetName()

Version        Name
-------        ----
5.0.0.0        System.Runtime.CompilerServices.Unsafe

```

Basically:

- System.Text.Json 5.0.0.1 references System.Memory 4.0.1.1 which references System.Runtime.CompilerServices.Unsafe 4.0.4.1
- However, when I use `dotnet publish` with a classlib that references System.Text.Json, that pulls down System.Runtime.CompilerServices.Unsafe 5.0.0.0, which isn't compatible
- Because this is a PowerShell module, I can't write a static binding redirect to make this work

I'm going to try writing an `Assembly.Resolve` event handler for this particular assembly, and if successful I'll post the code here

**Comment by rjmholt** [comment]
So the only issue is that the failure occurs when we try to load System.Text.Json rather than with System.Runtime.CompilerServices.Unsafe directly, even though we have the right assembly version.

So I just wrote a generic dynamic binding redirect handler which isn't ideal, but gets the job done:

```csharp
#if !CoreCLR
    public class UnsafeCompilerServicesLoadHandler : IModuleAssemblyInitializer, IModuleAssemblyCleanup
    {
        private static readonly string s_moduleAsmDir = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);

        public void OnImport()
        {
            AppDomain.CurrentDomain.AssemblyResolve += HandleAssemblyResolve;
        }

        public void OnRemove(PSModuleInfo psModuleInfo)
        {
            AppDomain.CurrentDomain.AssemblyResolve -= HandleAssemblyResolve;
        }

        private static Assembly HandleAssemblyResolve(object sender, ResolveEventArgs args)
        {
            var requiredAsmName = new AssemblyName(args.Name);

            string possibleAsmPath = Path.Combine(s_moduleAsmDir, $"{requiredAsmName.Name}.dll");

            AssemblyName bundledAsmName = null;
            try
            {
                bundledAsmName = AssemblyName.GetAssemblyName(possibleAsmPath);
            }
            catch
            {
                // If we don't bundle the assembly we're looking for, we don't have it so return nothing
                return null;
            }

            // Now make sure our version is greater
            if (bundledAsmName.Version < requiredAsmName.Version)
            {
                return null;
            }

            return Assembly.LoadFrom(possibleAsmPath);
        }
    }
#endif
```

**Comment by ericstj** [comment]
Be careful using AssemblyResolve handler.  Your handler only works if the load fails.  The load might succeed and fail to redirect if the exact version is in the GAC, or on one of PowerShell's probing paths, or already loaded in the AppDomain: which can cause type unification issues.

> Because this is a PowerShell module, I can't write a static binding redirect to make this work

I thought PowerShell supported the `.dll.config` convention for modules, creating an appdomain for those modules and loading their own config-file.

**Comment by rjmholt** [comment]
> I thought PowerShell supported the .dll.config convention for modules, creating an appdomain for those modules and loading their own config-file.

I've never heard of that, but I haven't tried it to confirm. My understanding was that modules are all loaded into the same appdomain with `Assembly.LoadFrom()`. I haven't heard of anyone using a `.dll.config` file -- is that a wider phenomenon in .NET Framework?

**Comment by rjmholt** [comment]
> Be careful using AssemblyResolve handler. Your handler only works if the load fails. The load might succeed and fail to redirect if the exact version is in the GAC, or on one of PowerShell's probing paths, or already loaded in the AppDomain: which can cause type unification issues.

Yeah that's why I didn't want to go down that route, it was just a last resort.

**Comment by ericstj** [comment]
Using `.dll.config` is a convention that's used by many .NETFramework plugin hosts to permit plugins to describe their bindingRedirects.  I know must test-runners (MSTest, XUnit, etc) use this convention, Azure PAAS used it, and *I thought* powershell used it to, but I can't find the docs on it.  Maybe @SteveL-MSFT would remember.
You could always implement that your self in a proxy component, but it would have some amount of overhead.

**Comment by hawkhsieh** [comment]
> Also facing similar problem.
> In my scenario, I've built a server-client structure application.
> 
> ```
> CommonLibrary: .netstandard2.0 (System.Text.Json-4.7.2 from nuget)
> Server: .NetCore3.1 for both core and wpf project (support System.Text.Json, internall)
> Client: .NetStandard2.0 for core and .NetFramework 4.7.2 fot wpf project (System.Text.Json-4.7.2 from nuget)
> ```
> 
> When I use something like
> `<AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>`
> and manually add
> 
> ```
>       <dependentAssembly>
>         <assemblyIdentity name="System.ValueTuple" publicKeyToken="cc7b13ffcd2ddd51" culture="neutral" />
>         <bindingRedirect oldVersion="0.0.0.0-4.0.3.0" newVersion="4.0.3.0" />
>       </dependentAssembly>
>       <dependentAssembly>
>         <assemblyIdentity name="System.Buffers" publicKeyToken="cc7b13ffcd2ddd51" culture="neutral" />
>         <bindingRedirect oldVersion="0.0.0.0-4.0.3.0" newVersion="4.0.3.0" />
>       </dependentAssembly>
> ```
> 
> , the client applications still throw same exception
> 
> ```
> System.IO.FileLoadException: 未能加载文件或程序集“System.Buffers, Version=4.0.2.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51”或它的某一个依赖项。找到的程序集清单定义与程序集引用不匹配。 (异常来自 HRESULT:0x80131040)
> ```
> 
> To reduce the complexity of the application, I've change all System.Text.Json code back to Newtonsoft.Json.
> Hope there is a more proper way to resolve this issue

This is a convenient solution! 
I also fix this problem.
